[MSE][GStreamer] Prevent currentTime reset on preroll when audio sink is non-async

eocanha opened the pull request:
#### 1791f0d769bbbad7317d9f8ac7f89b1957d9a217
<pre>
[MSE][GStreamer] Prevent currentTime reset on preroll when audio sink is non-async
<a href="https://bugs.webkit.org/show_bug.cgi?id=269587">https://bugs.webkit.org/show_bug.cgi?id=269587</a>

Reviewed by Xabier Rodriguez-Calvar.

When doing an initial seek (during preroll) on an MSE content that is
audio-only on platforms whose audio sink is not asynchronous (eg: on
Broadcom Nexus devices), the manually triggered preroll introduced by
<a href="https://commits.webkit.org/268845@main">https://commits.webkit.org/268845@main</a> causes an invalidation of the
currentTime cached position. This causes a temporary currentTime 0 value
that creates a drop to HaveMetadata in the MediaSource readyState,
causing a number of problems.

See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1211">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1211</a>

This patch remembers seek position as a cached value on an initial seek
during preroll when the audio sink can&apos;t perform asynchronous state
changes.

Based on a patch by Jacek Skiba &lt;jacek.skiba@consult.red&gt;

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::doSeek): Cache the previous position (the seek position) before the manual preroll and then immediately propagate the readyState to the player and invalidate the cached position after the preroll, so the new position is the one resulting from the preroll to the seek target position.

Canonical link: <a href="https://commits.webkit.org/274971@main">https://commits.webkit.org/274971@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/74be3394eb23badbd370c96b730da5bc30ba5614

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40194 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19206 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42572 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42739 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36283 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42501 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22129 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16535 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33413 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40768 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16179 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34710 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13982 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14023 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35662 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44017 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36460 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35991 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39739 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15009 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12310 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38016 "Found 1 new API test failure: /WebKitGTK/TestWebKitUserContentFilterStore:/webkit/WebKitUserContentFilterStore/filter-persistence (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16628 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9083 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16677 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16272 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris reviewed (approved):
Please, ensure there are no regressions on tests.

eocanha committed:
[MSE][GStreamer] Prevent currentTime reset on preroll when audio sink is non-async
https://bugs.webkit.org/show_bug.cgi?id=269587

Reviewed by Xabier Rodriguez-Calvar.

When doing an initial seek (during preroll) on an MSE content that is
audio-only on platforms whose audio sink is not asynchronous (eg: on
Broadcom Nexus devices), the manually triggered preroll introduced by
https://commits.webkit.org/268845@main causes an invalidation of the
currentTime cached position. This causes a temporary currentTime 0 value
that creates a drop to HaveMetadata in the MediaSource readyState,
causing a number of problems.

See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/1211

This patch remembers seek position as a cached value on an initial seek
during preroll when the audio sink can't perform asynchronous state
changes.

Based on a patch by Jacek Skiba <jacek.skiba@consult.red>

* Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp:
(WebCore::MediaPlayerPrivateGStreamerMSE::doSeek): Cache the previous position (the seek position) before the manual preroll and then immediately propagate the readyState to the player and invalidate the cached position after the preroll, so the new position is the one resulting from the preroll to the seek target position.

Canonical link: https://commits.webkit.org/274971@main
